### DISCUSSION TITLE
Basic IOS support

### DISCUSSION THREAD

**Discussion by BrendanSimon** [hbldh/bleak#1248] [Q&A]
I want to write an app for iOS and Android that communicates with a Linux device to obtain some basic status information and display it in the app.

The linux devices may have Ethernet, serial console, WiFi or Bluetooth available (e.g. via USB adapters).  Since the devices are installed high off the ground I was thinking a wireless technology would be better (and better suits mobile phones as they have WiFi and Bluetooth built-in).  Bluetooth seems like the better option as WiFi connection might mean losing Internet connection (e.g. if WiFi hot spotting).

My intention was to use Beeware Toga for the GUI side of the app.  As far as I'm aware Toga doesn't have support for Bluetooth services, so this is where BLEAK could play a part, but iOS support seems to be lacking.

**Is there any work going on in the BLEAK community to provide basic iOS support?**

I think I'd be happy with just a serial link over Bluetooth so that an app on Linux could just send out some text status info which the Toga app could parse and display.

I am open to other suggestions too as I am in the brainstorming phase.

Cheers :)

**Comment by dlech** [comment]
There aren't any plans to add iOS support to Bleak. Since iOS is nearly the same as macOS, it would probably be fairly trivial to port the existing PyObjC for macOS backend to rubicon-objc. But I have a feeling that it would end up unmaintained like the Android backend currently is (you can't just run a script on Android for a quick test, you have to build a full app and compile the entire Python runtime, so it just takes too much time).

**Reply by BrendanSimon** [reply]
I haven't done any Android work at all, but from the snippets I've read over the years I also get the feeling it's a big process :(

I did play with Toga and deploying an app to run/test in the xcode simulator and that wasn't overly painful.  I'm sure I didn't have to build python (not even once), so my gut feel is that the python side of things is not an issue.  If BLEAK is pure python that wraps the underlying OS libraries then I think there's less chance of things rusting and being unmaintainable.

I note that Beeware and Toga now have an automatic test framework in place for most platforms, including iOS and Android.  Coverage is very minimal at this stage, but the frameworks are there.  I'm not sure if any of that can be used with BLEAK (or vice versa) to keep various targets maintained.  They also mention improvements for Android build times in their latest update, which may be worth a look.

https://beeware.org/news/buzz/january-2023-status-update/
https://beeware.org/news/buzz/february-2023-status-update/
https://beeware.org/news/

On another note, could you mock the calls to Android library calls so that you run your tests on dev platforms, etc?  At least you could verify that various calls are made with expected parameter values, and assume Android does the right thing from there.

**Reply by MarkusPiotrowski** [reply]
I also want to confirm that with the beeware framework it's quite easy and fast to make a small Android apk for testing. I'm myself much interested in having BLE support for Android. We tried to modify the existing python4android backend of bleak for using it in beeware (with the help of the chaquopy developer) here: https://github.com/beeware/beeware/issues/181 but we failed so far.

**Reply by dlech** [reply]
FYI, I'm available for hire and have some availability currently if anyone needs commercial support for adding support for BeeWare.
